sdk/identity/azure-identity/azure/identity/_credentials/app_service.py

```diff
@@ -0,0 +1,113 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import functools
+import os
+from typing import TYPE_CHECKING
+
+from azure.core.credentials import AccessToken
+from azure.core.pipeline.transport import HttpRequest
+
+from .. import CredentialUnavailableError
+from .._constants import EnvironmentVariables
+from .._internal.managed_identity_client import ManagedIdentityClient
+from .._internal.get_token_mixin import GetTokenMixin
+
+if TYPE_CHECKING:
+    from typing import Any, Optional
+
+
+class AppServiceCredential(GetTokenMixin):
+    def __init__(self, **kwargs):
+        # type: (**Any) -> None
+        super(AppServiceCredential, self).__init__()
+
+        client_args = _get_client_args(**kwargs)
+        if client_args:
+            self._client = ManagedIdentityClient(**client_args)
+        else:
+            self._client = None
+
+    def get_token(self, *scopes, **kwargs):
+        # type: (*str, **Any) -> AccessToken
+        if not self._client:
+            raise CredentialUnavailableError(
+                message="App Service managed identity configuration not found in environment"
+            )
+        return super(AppServiceCredential, self).get_token(*scopes, **kwargs)
+
+    def _acquire_token_silently(self, *scopes):
+        # type: (*str) -> Optional[AccessToken]
+        return self._client.get_cached_token(*scopes)
+
+    def _request_token(self, *scopes, **kwargs):
+        # type: (*str, **Any) -> AccessToken
+        return self._client.request_token(*scopes, **kwargs)
+
+
+def _get_client_args(**kwargs):
+    # type: (dict) -> Optional[dict]
+    identity_config = kwargs.pop("_identity_config", None) or {}
+
+    url = os.environ.get(EnvironmentVariables.IDENTITY_ENDPOINT)
+    secret = os.environ.get(EnvironmentVariables.IDENTITY_HEADER)
+    if url and secret:
+        version = "2019-08-01"
+        base_headers = {"X-IDENTITY-HEADER": secret}
+        content_callback = None
+    else:
+        url = os.environ.get(EnvironmentVariables.MSI_ENDPOINT)
```

Comment:
Is it possible that IDENTITY_ENDPOINT, IDENTITY_HEADER, MSI_ENDPOINT, MSI_SECRET all exist?

And what's the behavior for this case?

Reply:
Yes, in fact that's to be expected because App Service instances which support the new version still support the old one for compatibility's sake. Behavior is what you see here: prefer the more recent version, fall back to the prior one if necessary. The fallback behavior is required because the new API isn't supported for all apps, for example Linux Functions apps on a consumption plan don't support it.

Reply:
If all of them exist, will we try "2017-09-01" APIs if "2019-08-01" does not work?

Reply:
Not sure I understand your question but if you're asking whether we will follow a failed request to version 2019-08-01 with a request to 2017-09-01, no, we won't.

Reply:
Yes. You understand my question correctly. My question was do we want to make it behave like chained credential. if "2019-08-01" cannot get a valid token, we will try "2017-09-01". Thanks for the answer.